demux/demux_lavf.c

```diff
@@ -1150,25 +1150,22 @@ static int demux_open_lavf(demuxer_t *demuxer, enum demux_check check)
     demuxer->is_network |= priv->format_hack.is_network;
     demuxer->seekable &= !priv->format_hack.no_seek;
 
-    if (priv->avfc->duration > 0) {
-        demuxer->duration = (double)priv->avfc->duration / AV_TIME_BASE;
-    } else {
-        double total_duration = 0;
-        double av_duration = 0;
-        for (int n = 0; n < priv->avfc->nb_streams; n++) {
-            AVStream *st = priv->avfc->streams[n];
-            if (st->duration <= 0)
-                continue;
-            double f_duration = st->duration * av_q2d(st->time_base);
-            total_duration = MPMAX(total_duration, f_duration);
-            if (st->codecpar->codec_type == AVMEDIA_TYPE_AUDIO ||
-                st->codecpar->codec_type == AVMEDIA_TYPE_VIDEO)
-                av_duration = MPMAX(av_duration, f_duration);
-        }
-        double duration = av_duration > 0 ? av_duration : total_duration;
-        if (duration > 0)
-            demuxer->duration = duration;
+    double total_duration = 0;
+    double av_duration = 0;
+    for (int n = 0; n < priv->avfc->nb_streams; n++) {
+        AVStream *st = priv->avfc->streams[n];
+        if (st->duration <= 0)
+            continue;
+        double f_duration = st->duration * av_q2d(st->time_base);
+        total_duration = MPMAX(total_duration, f_duration);
+        if (st->codecpar->codec_type == AVMEDIA_TYPE_AUDIO ||
+            st->codecpar->codec_type == AVMEDIA_TYPE_VIDEO)
+            av_duration = MPMAX(av_duration, f_duration);
     }
+    double duration = av_duration > 0 ? av_duration : total_duration;
+    if (duration < 0 && priv->avfc->duration > 0)
```

Comment:
is duration actually ever < 0?
I think don't we're meant to prefer a zero detected duration over a non-zero duration from avfc.

Reply:
The duration is -1 if ffmpeg couldn't detect it with the per-stream method. 0 detected duration is just image files

Reply:
Nope, this is wrong. For image files, you get -Int64 not -1.

And we discard those in `if (st->duration <= 0) continue`. Currently, mpv sees a duration of 0 for image files because avfc duration is -Int64, and we discard the -Int64 from `st->duration` and return 0.

And duration should never be zero since ffmpeg will calculate it via the bitrate and filesize if the stream header doesn't specify it.

total_duration of 0 is valid and duration is never actually < 0, so I think my original `duration <= 0` was more correct

Reply:
I made it `duration == 0 && priv->avfc->duration > 0` since duration can never actually be < 0